src/mlir/dialect/VeronaTypes.h

```diff
@@ -0,0 +1,108 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "mlir/IR/Dialect.h"
+
+namespace mlir::verona
+{
+  Type parseVeronaType(DialectAsmParser& parser);
+  void printVeronaType(Type type, DialectAsmPrinter& os);
+
+  /// Returns true if the type is one defined by the Verona dialect.
+  bool isaVeronaType(Type type);
+  /// Returns true if all types in the array are ones defined by the Verona
+  /// dialect.
+  bool areVeronaTypes(llvm::ArrayRef<Type> types);
+
+  namespace detail
+  {
+    struct MeetTypeStorage;
+    struct JoinTypeStorage;
+    struct IntegerTypeStorage;
+    struct CapabilityTypeStorage;
+  }
+
+  // In the long term we should claim a range in LLVM's DialectSymbolRegistry,
+  // rather than use the "experimental" range.
+  static constexpr unsigned FIRST_VERONA_TYPE =
+    Type::Kind::FIRST_PRIVATE_EXPERIMENTAL_0_TYPE;
+  static constexpr unsigned LAST_VERONA_TYPE =
+    Type::Kind::LAST_PRIVATE_EXPERIMENTAL_0_TYPE;
+
+  namespace VeronaTypes
+  {
+    enum Kind
+    {
+      Meet = FIRST_VERONA_TYPE,
+      Join,
+      Integer,
+      Capability,
+    };
+  }
+
+  struct MeetType
+  : public Type::TypeBase<MeetType, Type, detail::MeetTypeStorage>
+  {
+    using Base::Base;
+    static MeetType
+    get(MLIRContext* ctx, llvm::ArrayRef<mlir::Type> elementTypes);
+    llvm::ArrayRef<mlir::Type> getElements() const;
+
+    static bool kindof(unsigned kind)
+    {
+      return kind == VeronaTypes::Meet;
+    }
+  };
+
+  struct JoinType
+  : public Type::TypeBase<JoinType, Type, detail::JoinTypeStorage>
+  {
+    using Base::Base;
+    static JoinType
+    get(MLIRContext* ctx, llvm::ArrayRef<mlir::Type> elementTypes);
+    llvm::ArrayRef<mlir::Type> getElements() const;
+
+    static bool kindof(unsigned kind)
+    {
+      return kind == VeronaTypes::Join;
+    }
+  };
+
+  struct IntegerType
+  : public Type::TypeBase<IntegerType, Type, detail::IntegerTypeStorage>
+  {
+    using Base::Base;
+
+    static IntegerType get(MLIRContext* context, size_t width, unsigned sign);
+
+    size_t getWidth() const;
+    bool getSign() const;
+
+    static bool kindof(unsigned kind)
+    {
+      return kind == VeronaTypes::Integer;
+    }
+  };
+
+  enum class Capability
```

Comment:
Why is this enum here and not inside the storage class, like Integers?

Reply:
Unlike Integer's `SignType` (which ends up being exposed as a bool anyway), the capability inside the type is very much part of its public interface. In various places during typechecking we'll need to look at the capability to make decisions.

Reply:
Integer sign and width are also part of their public interfaces, that you expose using `getImpl()`. What's the difference?

Reply:
The sign is exposed as a boolean, so the definition of its enum, `SignType`, doesn't need to be public.
For capability, we definitely want to expose it as an enum, not as a (set of) boolean.

Reply:
What about integer width? That's not a boolean either.

Reply:
I'm sorry I don't understand your point.

You were asking why the `Capability` enum is defined in the header and not in the source file, as part of the storage class. The answer is because it is part of method signatures of `CapabilityType` (both `get` and `getCapability`). Therefore it needs to be exposed to everyone that uses `CapabilityType`, hence must be in the header.

All methods on `IntegerType` use standard C++ types. The sign is exposed as a boolean and the width as an integer. `SignType` is an implementation detail of the storage class, hence it can happily live in the `.cpp` file.

Reply:
D'oh! Now I get it, sorry. Friday.